src/display_context.cpp

```diff
@@ -87,21 +87,60 @@ display_context::can_move_result display_context::unit_can_move(const unit& u) c
 	const team& current_team = get_team(u.side());
 
 	can_move_result result = {false, false};
-	for(const map_location& adj : get_adjacent_tiles(u.get_location())) {
-		if (map().on_board(adj)) {
-			if(!result.attack_here) {
-				const unit_map::const_iterator i = units().find(adj);
-				if (i.valid() && !i->incapacitated() && current_team.is_enemy(i->side())) {
-					result.attack_here = true;
+
+	// Generating pairs of min-max ranges of all units weapons
+	const auto& attacks = u.attacks();
+	std::vector<std::pair<int, int>> distance_ranges;
+	for (const auto& attack : attacks) {
+		int min_range = attack.min_range();
+		int max_range = attack.max_range();
+		distance_ranges.emplace_back(min_range , max_range);
+	}
+
+	// Merging it here, may be removed if some checks for every weapon 
+	// Should be done individually (Like line-of sight)
+	std::sort(distance_ranges.begin(), distance_ranges.end());
+
+	std::vector<std::pair<int, int>> merged_ranges;
+	for (const auto& range : distance_ranges) {
+		if (merged_ranges.empty() || merged_ranges.back().second < range.first) {
+			merged_ranges.push_back(range);
+		} else {
+			merged_ranges.back().second = std::max(merged_ranges.back().second, range.second);
+		}
+	}
+
+
+	for (const auto& range : merged_ranges) {
+		int min_distance = range.first;
+		int max_distance = range.second;
+
+		for (int dx = -max_distance; dx <= max_distance; ++dx) {
+			for (int dy = -max_distance; dy <= max_distance && !result.attack_here; ++dy) {
+				// Adjust for hex grid
+				int adjusted_dy = dy + (dx - (dx&1)) / 2;
```

Comment:
You're right... idk how to do this?
(there's no function and I used similar thing in Lua (copied from some other code while ago)... and it just worked)
...
Any idea how to the iteration then?

Reply:
I'm not even sure that what you're doing is wrong, so no, I don't have any idea of how to do it better. It just looks strange to be looking at a rectangle when what you want is a circle.

Reply:
Found no faster way, the over-locations aren't too big generally it's about a 1/4 of all test cases (tested on numbers from 1 to 5).
Is this acceptable?

Reply:
What does `dx&1` do?

Reply:
Yeah, is it possible to do this without introducing a bitwise operator? As CelticMinstrel earlier said, it probably isn't doing anything.

Reply:
From discussing on discord:
> Anyway, the formula `dy + (dx - (dx&1)) / 2` is exactly equivalent to `dy + dx / 2` AFAICT.

so please use that way - it's simpler and clearer what it's doing.

Reply:
> Yeah, is it possible to do this without introducing a bitwise operator? As CelticMinstrel earlier said, it probably isn't doing anything.

What do you mean, that it's doing anything? It can sub 1 or 0.

Reply:
It behaves differently when dx is negative, as then the `dx/2` rounds up.

Reply:
Okay, let me break down that formula.

`dy + (dx - (dx&1)) / 2`

First of all, as you say, `dx&1` is either 1 or 0 – specifically, it's 0 if dx is even and 1 if dx is odd. So let's treat these as two separate cases and see what happens.

**Case 1**: dx is even

In this case `dx&1` is 0, so it vanishes and we are left with `dy + dx / 2`.

**Case 2**: dx is odd

In this case, `dx&1` is 1, so subtracting that from dx gives us an even number. So we're left with `dy + dx / 2`.

This second case might not be obvious if you think of it in terms of ordinary mathematics, but the division operator you see here is _not conventional division_. It's truncating division. So 2/2 is 1, and 3/2 is also one. By extension, if dx is odd, then `(dx-1)/2` is equal to `dx/2`.

So in summary, your formula reduces to just `dy + dx / 2` in all cases.

…well, almost all cases. If dx is negative, then it won't be the same. For simplicity of demonstration, suppose that dy is 0. Now, if dx is 3, then both your formula and the simplified one will yield 1. But if dx is -3, then that's not the case. Your formula will produce -2, while the simplified one will produce -1. In more technical terms, you are doing a floor (round towards negative infinity), whereas the default operation upon division is truncation (round towards zero).

Is a negative dx expected here? Perhaps it is, I'm not quite sure. But even if it is, then the formula is still much more readable expressed as `dy + floor(dx / 2.0)` rather than with a bitwise operator. Note that the decimal point in `2.0` is required here to make division behave as proper division, rather than truncating division – `3 / 2` is 1, but `3 / 2.0` is 1.5. That said, there's also a possibility that this difference in the case of negative numbers was actually an error and it _should_ have been giving -1 in the case of dx=-3, dy=0. I don't know what's correct here.

You might argue that floor is less efficient that the bitwise operation, and technically, you'd probably be right. But unless you can prove (via proper profiling) that there is a noticeable performance boost from using the bitwise operation, there's no point in making the code unreadable in the name of performance. In fact, even if you _did_ prove there was a noticeable performance gain, the proper solution would be to write a custom floor:

```cpp
inline int floor(int n) { return n - (n & 1); }
```

Reply:
dx iterates from -max to max.
Why do I have to put an inline function there for one operation, that seems pointless, I can comment it in the code, if the readability is really the case there.

Reply:
Is max_range/min_range really supposed to be negative? What does a negative range do? (it's stored in an `int` but has default value of 1, so perhaps that should've been `unsigned`?)

Reply:
I'm not saying you have to put an inline function. I'm saying your code should be understandable at a glance.

If the intent is to calculate `dy + dx / 2`, where division is truncating division, then just write it like that.

If the intent is to calculate `dy + floor(dx / 2.0)`, where division is regular floating-point division, then write it like that.

The example floor function I wrote is not necessary. (Actually, it's not even correct.) It was just an example of what you'd do if you needed `dy + floor(dx / 2.0)` but found via profiling that it's unacceptably slow… which is to say, it's something that you shouldn't be thinking of right now.

---

So, the question is, what do you want this formula to produce when you have dy = 0 and dx = -3?

If `adjusted_dy` should be -1 in this case, then write the formula as `dy + dx / 2`.

On the other hand, if `adjusted_dy` should  be -2 in this case, then write the formula as `dy + floor(dx / 2.0)`.

Reply:
Haven't tested.
But dx and dy is ofc supposed to be negative.

Reply:
> Is max_range/min_range really supposed to be negative? What does a negative range do? (it's stored in an `int` but has default value of 1, so perhaps that should've been `unsigned`?)

Haven't tested.
But dx and dy is ofc supposed to be negative.

Reply:
If it works correctly with the bitwise operator when dx is negative, then you probably need the version with `floor`. If it gives wrong answers with the bitwise operator when dx is negative, then you need the simpler version without `floor`.

Reply:
> On the other hand, if `adjusted_dy` should  be -2 in this case, then write the formula as `dy + floor(dx / 2.0)`.
If I can't use the &, then that's the best option I guess.

Reply:
done